px4_ros2_cpp/src/components/message_compatibility_check.cpp

```
@@ -229,89 +350,116 @@ bool messageCompatibilityCheck(
-      auto last_slash = message_to_check.topic_name.find_last_of('/');
-      if (last_slash == std::string::npos) {
-        topic_type = message_to_check.topic_name;
-      } else {
-        topic_type = message_to_check.topic_name.substr(last_slash + 1);
-      }
-      topic_type = snakeToCamelCase(topic_type);
-    }
-    // Read the local message definition and get the hash
-    const uint32_t expected_message_hash = messageHash(node, topic_type, msgs_dir);
-
-    // Ask for the message hash from PX4
-    px4_msgs::msg::MessageFormatRequest request;
-    request.protocol_version = px4_msgs::msg::MessageFormatRequest::LATEST_PROTOCOL_VERSION;
-    strncpy(
-      reinterpret_cast<char *>(request.topic_name.data()),
-      message_to_check.topic_name.c_str(), request.topic_name.size() - 1);
-    request.topic_name.back() = '\0';
-    request.timestamp = 0; // Let PX4 set the timestamp
-    px4_msgs::msg::MessageFormatResponse response;
-    switch (requestMessageFormat(
-        node, request, message_format_response_sub, message_format_request_pub,
-        response, first_message))
-    {
-      case RequestMessageFormatReturn::Timeout:
-        RCLCPP_FATAL(
-          node.get_logger(),
-          "Timed out waiting for message format. Is the FMU running?");
-        // Do not try to check the other formats
-        return false;
-      case RequestMessageFormatReturn::ProtocolVersionMismatch:
-        // Error already reported
-        return false;
-      case RequestMessageFormatReturn::GotReply:
-        if (response.success) {
-          if (response.message_hash != expected_message_hash) {
-            mismatched_topics += "\n  - " + message_to_check.topic_name;
+
+  if(px4_ros2::is_rmw_zenoh()) {
+    for (const auto & entry : all_px4_ros2_messages) {
+      auto publishers_info = node.get_publishers_info_by_topic(entry.topic_name);
+      const rosidl_type_hash_t *pub_hash;
+
+      if (!publishers_info.empty()) {
```

So if not advertised yet this will silently succeed (?)

---

Well if there's no publisher advertising it, there's no way to check if the hashes match.
Unless there should also be a check if all these topics are being published?

---

I don't have a better solution, just that this does not provide any guarantee. The minimum we can do is document it in the API docs.